### PR TITLE
Use Miller-Rabin primality testing

### PR DESCRIPTION
**Pull request by adamantike**
Use **_Miller-Rabin_ primality testing** **[1]**, instead of Solovay-Strassen primality testing (currenly used). It reduces theoretical error probability to ```4^(-k)``` (Solovay-Strassen is ```2^(-k)```).

**Miller-Rabin** is required by NIST FIPS 186-4 **[2]** as probabilistic primality test (Appendix C, subsection C.3):

> This Standard allows two alternatives for testing primality: either using several iterations of only the Miller-Rabin test, or using the iterated Miller-Rabin test, followed by a single Lucas test.

Deprecated functions since this PR:
* ```jacobi()```
* ```jacobi_witness()```
* ```randomized_primality_testing()```

These functions can be removed, as they are Solovay-Strassen's implementation.

**[1]** https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test
**[2]** http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pd

### CODE REVIEW

**Comment by adamantike** [comment]
First commit implements Miller-Rabin.
Second commit removes Solovay-Strassen implementation.

File: rsa/prime.py [hunk 1]
```diff
@@ -103,13 +90,32 @@ def randomized_primality_testing(n, k):
...
     >>> is_prime(41)
     True
+    >>> [x for x in range(901, 1000) if is_prime(x)]
+    [907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]
     """
 
-    return randomized_primality_testing(number, 6)
+    # Check for small numbers.
+    if number < 10:
+        return number in [2, 3, 5, 7]
```
Comment (sybrenstuvel) [inline rsa/prime.py]: Never build a list for this. Using a set literal is 33% faster (on my simple test) -- the percentage can change, but a set test is faster than a list test. I'll handle this in another commit.

File: rsa/prime.py [hunk 2]
```diff
@@ -103,13 +90,32 @@ def randomized_primality_testing(n, k):
...
+        return False
+
+    # According to NIST FIPS 186-4, Appendix C, Table C.3, minimum number of
+    # rounds of M-R testing, using an error probability of 2 ** (-100), for
+    # different p, q bitsizes are:
+    #   * p, q bitsize: 512; rounds: 7
+    #   * p, q bitsize: 1024; rounds: 4
+    #   * p, q bitsize: 1536; rounds: 3
+    # See: http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf
+    return miller_rabin_primality_testing(number, 7)
```
Comment (sybrenstuvel) [inline rsa/prime.py]: Do you think we can let `k` depend on the bitsize of `number`? That might make things faster for larger numbers.

File: rsa/prime.py [hunk 3]
```diff
@@ -103,13 +90,32 @@ def randomized_primality_testing(n, k):
 def is_prime(number):
     """Returns True if the number is prime, and False otherwise.
 
+    >>> is_prime(2)
+    True
     >>> is_prime(42)
     False
     >>> is_prime(41)
     True
+    >>> [x for x in range(901, 1000) if is_prime(x)]
```
Comment (sybrenstuvel) [inline rsa/prime.py]: This should be moved to a unit test, as it doesn't show any extra information for the user.

**Comment by sybrenstuvel** [comment]
There are some issues with the code. Sometimes unit tests hang indefinitely. I think it has something to do with the use of unchecked while-loops. This:

        # Generate random integer a, where 2 <= a <= (n - 2)
        a = 0
        while a < 2:
            a = rsa.randnum.randint(n - 2)

can be written without a loop, like this:

            a = rsa.randnum.randint(n - 4) + 2

This loop also was unchecked, and would hang indefinitely when `n = 1`:

    d = n - 1
    r = 0
    while not (d & 1):
        r += 1
        d >>= 1

I'll fix this in another commit, but of course please take this into consideration in future pull requests.

**Comment by sybrenstuvel** [comment]
Merged in 6ac63ad2998be9f4f8ed84dd671fba1811be7256

**Comment by adamantike** [comment]
@sybrenstuvel I will gladly fix my PRs based on your reviews, before you merge them!

**Comment by sybrenstuvel** [comment]
I know, I just wanted to get 3.4 out of the door today (Thursday is my day off work). The commit is still under your name, so you get all the credit ;-)

**Comment by adamantike** [comment]
Not really interest in credit, but to really take advantage of your reviews and improve my coding. Also, don't want to drain your free time by doing fixes I could do myself (you already took the time to review it and test it!) :)

**Comment by sybrenstuvel** [comment]
Excellent :)

**Pull request closed without merging**